src/core/lib/resource_quota/memory_quota.h

```diff
@@ -0,0 +1,322 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_RESOURCE_QUOTA_MEMORY_QUOTA_H
+#define GRPC_CORE_LIB_RESOURCE_QUOTA_MEMORY_QUOTA_H
+
+#include <grpc/support/port_platform.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <queue>
+#include <vector>
+
+#include "src/core/lib/gprpp/dual_ref_counted.h"
+#include "src/core/lib/gprpp/orphanable.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/promise/activity.h"
+#include "src/core/lib/promise/poll.h"
+
+namespace grpc_core {
+
+class Reclaimer;
+class MemoryAllocator;
+class MemoryQuota;
+
+// Reclamation passes.
+// When memory is tight, we start trying to claim some back from memory
+// reclaimers. We do this in multiple passes: if there is a less destructive
+// operation available, we do that, otherwise we do something more destructive.
+enum class ReclamationPass {
+  // Non-empty reclamation ought to take index 0, but to simplify API we don't
+  // expose that publicly (it's an internal detail), and hence index zero is
+  // here unnamed.
+
+  // Benign reclamation is intended for reclamation steps that are not
+  // observable outside of gRPC (besides maybe causing an increase in CPU
+  // usage).
+  // Examples of such reclamation would be resizing buffers to fit the current
+  // load needs, rather than whatever was the peak usage requirement.
+  kBenign = 1,
+  // Idle reclamation is intended for reclamation steps that are observable
+  // outside of gRPC, but do not cause application work to be lost.
+  // Examples of such reclamation would be dropping channels that are not being
+  // used.
+  kIdle = 2,
+  // Destructive reclamation is our last resort, and is these reclamations are
+  // allowed to drop work - such as cancelling in flight requests.
+  kDestructive = 3,
+};
+static constexpr size_t kNumReclamationPasses = 4;
```

Comment:
should this be 3?

Reply:
No... See the note on line 44... Need to think how to lay this out better.